proposals/0010-vk-buffer-ref.md

```diff
@@ -0,0 +1,270 @@
+
+The goal of this proposal is to have a solution that meets the following requirements:
+
+* Removes the need for having to manually or automatically generate offsets to load structured data with BufferDeviceAddress.
+* Get equivalent tooling functionality as is provided by the buffer reference feature in GLSL.  Namely, tools like RenderDoc are able to introspect the type information such that its buffer inspection and shader debugger are able to properly understand and represent the type of the data.
+* Make it possible through SPIR-V reflection to determine which members of a struct accessed by BufferDeviceAddress are statically referenced and at what offset.  This is already possible for other data like cbuffers in order for shader tooling to be able to identify which elements are used and where to put them.
+
+## Proposed solution
+
+Our solution is to add a new builtin type in the vk namespace that is a pointer to a buffer of a given type, `vk::buffer_ptr<T,A>`. The template argument `T` must be a struct. `A` must be an integer and is the alignment in bytes of the pointer. If `A` is not specified, the alignment is assumed to be 16 bytes.
+
+This new type will have the following operations
+
+* Copy assignment and copy construction - These copy the value of the pointer from one variable to another.
+* Accessing members of T - This is done using the `.` operator.
+* Conversion to uint64_t - Return the integer representation of the address it points to. This can be used for null pointer checks.
+
+Note the operations that are not allowed:
+
+* There is no default construction. Every vk::buffer_ptr<T> is either contained in a global resource (like a cbuffer, ubo, or ssbo), or it must be constructed using the copy constructor. 
+* There is no equivalent to the `*` operator that will dereference the entire struct.
+* There is no conversion from uint64_t to vk:buffer_ptr.
+* There is no explicit pointer arithmetic. All addressing is implicitly done using the `.` pointer, or indexing an array in the struct T.
+* The comparison operators == and != are not supported for buffer pointers. Comparison to null pointer should be done by first converting to uint64_t.
+
+Most of these restrictions are there for safety. They minimize the possibility of getting an invalid pointer. If the `.` operator is used on a null pointer, the behaviour is undefined.
+
+When used as a member in a buffer, vk::buffer_ptr can be used to pass physical buffer addresses into a shader, and address and access buffer space with logical addressing, which allows tools such as spirv-opt, spirv-reflect and renderdoc to be able to better work with these shaders.
+
+For example, here is a shader using vk::buffer_ptr to do the same thing as the shader above using vk::RawBufferLoad. Note the natural, logical syntax of the reference:
```

Comment:
to me all this describes a concept similar to a C++ reference, not a pointer. I suppose thats why its `GLSL_EXT_buffer_reference` and not `GLSL_EXT_buffer_pointer`.

Its not even about the the `.` being used over `->`

Not even the lack of pointer arithmetic.

The biggest thing that makes this a _reference_ and **not a pointer** is this
https://github.com/microsoft/DirectXShaderCompiler/pull/5089#issuecomment-1482072617
https://github.com/microsoft/DirectXShaderCompiler/pull/5089#issuecomment-1482087964
there seems to be no way to change the byte address in a struct, unless you somehow alias it with some other resource (which you obviously can't do for non-buffer variables because there's no `reinterpret_cast`) where the "pointer" is declared as a plain `uint64_t`.

Ergo my argument that this is a reference, and not a pointer.

Reply:
I think the real challenge here is that references, by definition, can't be null, and this does explicitly say it can be null.

I do agree with you that in some contexts this does behave more like a reference.

I can imagine a future where `vk::BufferPointer` could be implemented as something like a `std::optional<T &__attribute__((address_space(...) ))`, but it could also be more like a C++ smart pointer abstraction... I think it could go either way.

Reply:
I also find it to be more like a pointer because an assignment from one to another will change the address it points to and not the memory that it points to:

```
vk:buffer_ptr a = g1;
vk:buffer_ptr b = g2;
a.get() = 0;
b.get() = 1;
vk:buffer_ptr c = b;
c = a; // For a C++ reference you would expect b.get() to become 0, but in this case b.get() does not change.
```

Then as Chris mentioned, the ability for the value to be null makes it lean more towards a pointer for me.

Reply:
The example at the end with the linked list also makes me think pointer semantics.  In C terms, I'd think of it as a const pointer.